baseapp/state.go

```diff
@@ -17,5 +20,15 @@ func (st *state) CacheMultiStore() sdk.CacheMultiStore {
 
 // Context returns the Context of the state.
 func (st *state) Context() sdk.Context {
+	defer st.lock.RUnlock()
+	st.lock.RLock()
+
 	return st.ctx
```

Comment:
This is not actually safe, as the Context type has fields which have reference semantics.

Reply:
ouch

Reply:
@peterbourgon it seems the `Context` type needs some sort of deep copy.
thoughts

Reply:
tho more I dig into it - the more I am convinced this is a dirty hack and eventually it will blow
`BeginBlock` for example
if there are two servers listening (`socket` and `grpc`) they may (and will) call `BeginBlock` simultaneously.
correct if I'm getting it wrong

Reply:
Yep! I think you're right.

As far as I can see, BaseApp's exported methods — including but not limited to BeginBlock — can absolutely be called by concurrent goroutines. This means that they _must_ ensure that anything they read or write is synchronized. But that's not happening. BaseApp —and many, many, many other components in the SDK — permit unsynchronized reads and writes on their encapsulated values, and consequently violate Go's memory model. Many of these soundness errors remain undetected or overlooked because the current, specific execution paths _happen_ to not trigger them _most of the time_.

There are a lot of pathological issues — in both the Cosmos SDK generally, and the Context type specifically — which make these bugs hard to fix. 

In this case, at a high level: contexts are supposed to be request-scoped, but here — and many other places, too — the context is long-lived. I might be missing something, but that seems to be a clear design error. Neither a state nor a BaseApp nor anything else with a lifetime beyond an individual request should maintain a context value. 

More concretely: as with most types in the SDK, methods on the sdk.Context are — incorrectly — defined on a value receiver. That means every method call creates and operates on a (shallow) copy of the original value. This thrashes the GC and destroys performance, but, more importantly, it makes synchronization of any field with reference semantics more or less impossible. Even if you have a mutex in the type — which the context doesn't — those mutexes get copied, and so don't actually serve their purpose.

-- 

I certainly haven't done a deep-dive on this code, and so I'm not speaking from an informed place. But based on what I _do_ understand, it seems that the right approach to fixing this problem is to eliminate the `state` type altogether, including the deliverState and checkState fields in the BaseApp. Then, look at whatever was writing-to and reading-from those state values to figure out what stuff they actually needed from the contexts. Capture that information _specifically_, in a separate and synchronized type, in the BaseApp struct.

Or, I dunno. Maybe I got it all wrong.

Reply:
thanks @peterbourgon for looking over and confirming
I'll give it some thinking

Reply:
3. Construct a standalone app instance for quering which only shares the low-level db handler?

Reply:
That also might be an option, but im not sure it'll be an app. Rather We might have to refactor `Baseapp#Query`.

Reply:
@alexanderbez is this issue still a thing on later cosmos sdk 0.47+? or someone had opportunity to address it?